subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerCrossGroovyVersionIntegrationTest.groovy

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testkit.runner
+
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import org.gradle.util.GradleVersion
+import spock.lang.IgnoreIf
+
+@IgnoreIf({ GradleContextualExecuter.embedded })
+class GradleRunnerCrossGroovyVersionIntegrationTest extends BaseGradleRunnerIntegrationTest {
+
+    def "can run simple build using groovy2"(String testedGradleVersion, Closure<GradleRunner> gradleVersionConfigurator) {
+        given:
+        buildFile << """
+            task writeGroovyVersion {
+                doLast {
+                    file("version.txt").with {
+                        createNewFile()
+                        text = GroovySystem.version
+                    }
+                }
+            }
+        """
+
+        when:
+        def someRunner = runner().withArguments("writeGroovyVersion")
+        gradleVersionConfigurator.call(someRunner)
+        someRunner.build()
+
+        then:
+        file("version.txt").text == groovyVersion
+
+        cleanup:
+        testKitDaemons(GradleVersion.version(testedGradleVersion)).killAll()
+
+        where:
+        testedGradleVersion               | groovyVersion | gradleVersionConfigurator
+        "6.8.3"                           |  "2.5.12"     |  { it.withGradleVersion("6.8.3") }
+        GradleVersion.current().version   | "3.0.7"       |  { if (!GradleContextualExecuter.embedded) { it.withGradleInstallation(buildContext.gradleHomeDir) } }
+    }
+
+    def "groovy2-based gradle can run build via testkit using a groovy3-based gradle"() {
+        given:
+        def targetingGradle7Test = """
+import spock.lang.Specification
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+import org.gradle.testkit.runner.GradleRunner
+import org.gradle.testkit.runner.TaskOutcome
+import org.gradle.util.GradleVersion
+
+class BuildLogicFunctionalTest extends Specification {
+
+    @Rule public final TemporaryFolder testProjectDir = new TemporaryFolder();
```

Comment:
Locally when I run this test, the test-in-the-test is failing with `java.lang.IllegalAccessError: Update to non-static final field` (?) If I remove the `final` it works.

Reply:
Looks like this could be an issue with certain Java versions (I am using 11): https://github.com/spockframework/spock/issues/1011#issuecomment-582219001

We might run into this later in the pipeline as well.

Since this does not really matter for the test, maybe just remove the `final`.